mirage/stub/dns_stub_mirage.ml

```diff
@@ -0,0 +1,401 @@
+                        S.TCPV4.pp_error e) ;
+            Error (`Msg "connect failure")
+          | Ok flow ->
+            inc `Recursive_connections;
+            Lwt.async (fun () -> read_loop t flow);
+            t.flow <- Some flow;
+            Ok ({ t ; id = 0 })
+
+      let close _f =
+        (* ignoring this here *)
+        Lwt.return_unit
+
+      let recv { t ; id } =
+        let cond = Lwt_condition.create () in
+        t.requests <- IM.add id cond t.requests;
+        Lwt_condition.wait cond >|= fun data ->
+        t.requests <- IM.remove id t.requests;
+        match data with
+        | Ok cs -> Ok cs
+        | Error `Msg m -> Error (`Msg m)
+
+      let send f s =
+        match f.t.flow with
+        | None -> Lwt.return (Error (`Msg "no connection to resolver"))
+        | Some flow ->
+          let id = Cstruct.BE.get_uint16 s 2 in
+          f.id <- id;
+          S.TCPV4.write flow s >>= function
+          | Error e ->
+            f.t.flow <- None;
+            Lwt.return (Error (`Msg (Fmt.to_to_string S.TCPV4.pp_write_error e)))
+          | Ok () ->
+            inc `Recursive_queries;
+            Lwt.return (Ok ())
+    end
+
+    include Dns_client.Make(Transport)
+
+    let create ?size ?nameserver stack =
+      create ?size ~rng:R.generate ?nameserver ~clock:C.elapsed_ns stack
+  end
+
+  (* likely this should contain:
+     - a primary server (handling updates)
+     - a client on steroids: multiplexing on connections
+     - listening for DNS requests from clients:
+        first find them in primary server
+        if not authoritative, use the client
+  *)
+
+  (* task management
+     - multiple requests for the same name, type can be done at the same "time"
+     -> need to remember outstanding requests and signal to clients
+  *)
+
+  (* take multiple resolver IPs and round-robin / ask both (take first answer,
+     ignoring ServFail etc.) *)
+
+  (* timeout of resolver, retransmission (to another resolver / another flow) *)
+
+  module Dns_flow = Dns_mirage.Make(S)
+
+  type t = {
+    client : Client.t ;
+    reserved : Dns_server.t ;
+    mutable server : Dns_server.t ;
+    on_update : old:Dns_trie.t -> ?authenticated_key:[`raw] Domain_name.t -> update_source:Ipaddr.V4.t -> Dns_trie.t -> unit Lwt.t ;
+  }
+
+  let query_server trie question data build_reply =
+    match Dns_server.handle_question trie question with
+    | Ok (_flags, answer, additional) ->
+      (* TODO do sth with flags *)
+      inc `Authoritative_answers;
+      Some (build_reply ?additional (`Answer answer))
+    | Error (Rcode.NotAuth, _) -> None
+    | Error (rcode, answer) ->
+      inc `Authoritative_errors;
+      let data = `Rcode_error (rcode, Packet.opcode_data data, answer) in
+      Some (build_reply ?additional:None data)
+
+  let tsig_decode_sign server proto packet buf build_reply =
+    match Dns_server.handle_tsig server (Ptime.v (P.now_d_ps ())) packet buf with
```

Comment:
Two points:
- seems to me that `now` could be reused since there are no Lwt yield points here
- should we allow overriding this time with a `?time=(Ptime.v (P.now_d_ps ()))` parameter for the purpose of tests?

Reply:
fixed the first point; the second I'm not sure about -- developing a testsuite should lead to a refactoring anyways and hopefully establishing a (lwt-free) dns-stub package (atm it's all in dns-stub.mirage).

Reply:
fine with me :)